### PR TITLE
feat: [IDP-1437]: Implementation for Cookiecutter step

### PR DESCRIPTION
**Pull request by devesh-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55674)

### CODE REVIEW

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 1]
```diff
@@ -834,6 +843,41 @@ private static Map<String, String> getSecurityStepInfoEnvVariables(SecurityStepI
     return map;
   }
 
+  private static Map<String, String> getIDPCookieCuterStepInfoEnvVariables(
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:846]: Move to idp utils

**Comment by devesh-harness** [comment]
trigger uts

**Comment by devesh-harness** [comment]
trigger uts

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 27 of 30 files at r1, 4 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55674)* status: 33 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java` line 28 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlzyNsCed9WTikVqL5:-NjlzyNsCed9WTikVqL6:b-dox51a) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java#L28)):*
> ```Java
> import java.util.Map;
> 
> public class IDPStepUtils extends PluginServiceImpl {
> ```

nit: IDP Team



_Code quote:_
```Java
IDPStepUtils
```
___
*[`332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java` line 31 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njm-5vW7wxyqVvOEVgj:-Njm-5vW7wxyqVvOEVgk:b-1ts3tf) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java#L31)):*
> ```Java
>   public static Map<String, String> getIDPCookieCuterStepInfoEnvVariables(
>       IdpCookieCutterStepInfo stepInfo, String identifier) {
>     Map<String, String> map = new HashMap<>();
> ```

nit: we could rename it to what it represents. envMap or envVarMap in this case?



_Code quote:_
```Java
map
```
___
*[`332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java` line 44 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njm-yJb9tRKSNSfAIzl:-Njm-yJb9tRKSNSfAIzm:b-2tkqeo) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java#L44)):*
> ```Java
> 
>     PluginServiceImpl.setMandatoryEnvironmentVariable(map, IS_PUBLIC_TEMPLATE,
>         resolveStringParameter(
> ```

`resolveStringParameter` is deprecated. Can you check once if we need to use resolveStringParameterV2 instead?



_Code quote:_
```Java
resolveStringParameter
```
___
*[`idp-service/config/config.yml` line 403 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njlx5Bu10xDvtqj1O5H:-Njlx5Bu10xDvtqj1O5I:b-buzi81) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/config/config.yml#L403)):*
> ```YAML
>         - "C:\\harness-terraform.exe"
>     idpCookieCutter:
>       image: deveshharness2306/cookiecutter:latest
> ```

Is this for testing? Are we planning to have the image in some harness repo?



_Code quote:_
```YAML
deveshharness2306
```
___
*[`idp-service/src/main/java/io/harness/idp/pipeline/registrar/IdpStepRegistrar.java` line 46 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlxopQ8t35ftsZg1xH:-NjlxopQ8t35ftsZg1xI:b9khpql) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/pipeline/registrar/IdpStepRegistrar.java#L46)):*
> ```Java
>     engineSteps.put(BackgroundStep.STEP_TYPE, BackgroundStep.class);
>     engineSteps.put(RunTestsStep.STEP_TYPE, RunTestsStep.class);
>     engineSteps.put(IDPCookieCutterStep.STEP_TYPE, IDPCookieCutterStep.class);
> ```

Not related to this PR. Can we annotate this class with @UtilityClass?
___
*[`idp-service/src/main/java/io/harness/idp/pipeline/steps/IDPCookieCutterStep.java` line 16 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njly0v46oZ1a6og2iiK:-Njly0v46oZ1a6og2iiL:b-fxjbla) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/pipeline/steps/IDPCookieCutterStep.java#L16)):*
> ```Java
> 
> @OwnedBy(HarnessTeam.IDP)
> public class IDPCookieCutterStep extends AbstractStepExecutable {
> ```

Looks like we follow this naming convention IdpCookieCutterStep. Could we keep it consistent?



_Code quote:_
```Java
IDPCookieCutterStep
```
___
*[`idp-service/src/main/java/io/harness/idp/steps/Constants.java` line 12 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlyMznCxGsbYuFIBqA:-NjlyMznCxGsbYuFIBqB:bnpktju) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/Constants.java#L12)):*
> ```Java
> import io.harness.pms.contracts.steps.StepType;
> 
> public interface Constants {
> ```

Can we annotate this class with @UtilityClass?



_Code quote:_
```Java
Constants
```
___
*[`idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java` line 47 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlzQxm11W4SSxPH0zx:-NjlzQxm11W4SSxPH0zy:bynugzz) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java#L47)):*
> ```Java
> @Builder
> @AllArgsConstructor
> @RecasterAlias("io.harness.idp.pipeline.steps.beans.stepinfo.IdpCookieCutterStepInfo")
> ```

What is the right convention? For node, we have used stepNode (camelCase). Could we have the same convention in both places?



_Code quote:_
```Java
stepinfo
```
___
*[`idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java` line 59 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njlz70p0MBgDPHF_fVS:-Njlz70p0MBgDPHF_fVT:b-exvslt) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java#L59)):*
> ```Java
>   @Getter(onMethod_ = { @ApiModelProperty(hidden = true) }) @ApiModelProperty(hidden = true) private String name;
> 
>   @ApiModelProperty(dataType = STRING_CLASSPATH) ParameterField<String> isPublicTemplate;
> ```

Shouldn't this be Boolean?



_Code quote:_
```Java
@ApiModelProperty(dataType = STRING_CLASSPATH) ParameterField<String> isPublicTemplate;
```

**Review by devesh-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55674)* status: 33 file(s) to review. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java` line 44 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njm-yJb9tRKSNSfAIzl:-Njm5W0xCeM0YSYFz8V0:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/332-ci-manager/service/src/main/java/io/harness/idp/utils/IDPStepUtils.java#L44)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

`resolveStringParameter` is deprecated. Can you check once if we need to use resolveStringParameterV2 instead?
</blockquote></details>

done
___
*[`idp-service/config/config.yml` line 403 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njlx5Bu10xDvtqj1O5H:-Njm2Ovk51h8_31PI-VH:b-kzf3t8) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/config/config.yml#L403)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Is this for testing? Are we planning to have the image in some harness repo?
</blockquote></details>

yes, I will setup the separate repo and build a pipeline for the creation of these images, once that is done will replace it with images in the harness docker repo.
___
*[`idp-service/src/main/java/io/harness/idp/pipeline/registrar/IdpStepRegistrar.java` line 46 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlxopQ8t35ftsZg1xH:-Njm2pta7cXjz8bOfRff:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/pipeline/registrar/IdpStepRegistrar.java#L46)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Not related to this PR. Can we annotate this class with @UtilityClass?
</blockquote></details>

Done.
___
*[`idp-service/src/main/java/io/harness/idp/pipeline/steps/IDPCookieCutterStep.java` line 16 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njly0v46oZ1a6og2iiK:-Njm35HEAyYBfHfmYh2J:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/pipeline/steps/IDPCookieCutterStep.java#L16)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Looks like we follow this naming convention IdpCookieCutterStep. Could we keep it consistent?
</blockquote></details>

Done.
___
*[`idp-service/src/main/java/io/harness/idp/steps/Constants.java` line 12 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlyMznCxGsbYuFIBqA:-Njm3F3u8YwA2yMcO-5P:ben2fuj) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/Constants.java#L12)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we annotate this class with @UtilityClass?
</blockquote></details>

this is interface we cannot annotate with @UtilityClass
___
*[`idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java` line 47 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-NjlzQxm11W4SSxPH0zx:-Njm3Zrv7WCU3LHvaFvu:bk6poh1) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java#L47)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

What is the right convention? For node, we have used stepNode (camelCase). Could we have the same convention in both places?
</blockquote></details>

looks like in other places we have the same convention stepinfo
___
*[`idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java` line 59 at r3](https://reviewable.io/reviews/harness/harness-core/55674#-Njlz70p0MBgDPHF_fVS:-Njm3luKErfcXrCIL3Ss:b-ca3ggd) ([raw file](https://github.com/harness/harness-core/blob/851ac480101c352299d67338e62d72943ef9fa63/idp-service/src/main/java/io/harness/idp/steps/beans/stepinfo/IdpCookieCutterStepInfo.java#L59)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Shouldn't this be Boolean?
</blockquote></details>

no we cannot make it boolean as we discussed offline, its getting resolved as env variable.

**Review by sathish-soundarapandian — DISMISSED** [review]

**Review by vikyathharekal — DISMISSED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55674)* status: 33 file(s) to review. 1. unresolved discussions.

**Comment by devesh-harness** [comment]
trigger uts

**Review by harshjain12 — DISMISSED** [review]

**Comment by devesh-harness** [comment]
trigger uts

**Review by vikyathharekal — APPROVED** [review]

**Review by harshjain12 — APPROVED** [review]

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by sandeepa-kv-jois** [comment]
Unrelated test failures, going ahead with FM

**Pull request merged**